opampserver/pkg/connection/conncache.go

```diff
@@ -0,0 +1,144 @@
+package connection
+
+import (
+	"sync"
+	"time"
+
+	"github.com/odigos-io/odigos/common"
+	"github.com/odigos-io/odigos/opampserver/protobufs"
+	"google.golang.org/protobuf/proto"
+)
+
+const (
+	HeartbeatInterval = 30 * time.Second
+)
+
+// The time after which a connection is considered stale and can be cleaned up.
+var connectionStaleTime = time.Duration(float64(HeartbeatInterval) * 2.5)
+
+// Keep all live connections, with information about the connection.
+// The cache is cleaned up periodically to expire opamp clients that are no longer connected reporting data.
+type ConnectionsCache struct {
+	mux sync.Mutex
+
+	// map from device id to connection information
+	liveConnections map[string]*ConnectionInfo
+}
+
+func NewConnectionsCache() *ConnectionsCache {
+	return &ConnectionsCache{
+		liveConnections: make(map[string]*ConnectionInfo),
+	}
+}
+
+// GetConnection returns the connection information for the given device id.
+// the returned object is a by-value copy of the connection information, so it can be safely used.
+// To change something in the connection information, use the functions below which are synced and safe.
+func (c *ConnectionsCache) GetConnection(deviceId string) (*ConnectionInfo, bool) {
+	c.mux.Lock()
+	defer c.mux.Unlock()
+	conn, ok := c.liveConnections[deviceId]
+	if !ok || conn == nil {
+		return nil, false
+	} else {
+		// copy the conn object to avoid it being accessed concurrently
+		connCopy := *conn
+		return &connCopy, ok
+	}
+}
+
+func (c *ConnectionsCache) AddConnection(deviceId string, conn *ConnectionInfo) {
+	// copy the conn object to avoid it being accessed concurrently
+	connCopy := *conn
+	c.mux.Lock()
+	defer c.mux.Unlock()
+	c.liveConnections[deviceId] = &connCopy
```

Comment:
I remember we discussed about an option to save {deviceId}-{InstanceId} to handle fork cases?